Created api authorization middleware

jondeweydev opened the pull request:
### Summary
This middleware is in place so that the first time a user makes an API call to store project query data to a new project, their endpoint in use will be validated as well as their API key header.

### Type of Change
Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

### Evidence
-Unable to provide evidence currently, as no testing is currently in place.

jondeweydev committed:
created api authorization middleware

evanmcneely reviewed (commented):
Need to chat about how this fits into the rest of the logger functionality big picture. I get the intention, just want to know where your thinking is at.

jondeweydev committed:
fixed lint errors

jondeweydev committed:
added DB functionality for dev

jondeweydev committed:
fixed linting errors

jondeweydev committed:
fixed ts compilation error

jondeweydev committed:
move notes

jondeweydev committed:
removed DB configuration

jondeweydev committed:
added constructor URI param to keep const URI only in index

jondeweydev committed:
lint errors resolved

jondeweydev committed:
merge conflicts

jondeweydev committed:
jest configuration and skeleton for API verify/validate

jondeweydev committed:
testing updates

jondeweydev committed:
fixed lint errors

jondeweydev committed:
added logKey and projectID to constructor"

jondeweydev committed:
testing updates

jondeweydev committed:
lint error

jondeweydev committed:
api testing completed

jondeweydev commented:
Added Jest testing, all tests are passing.
Also removed express middleware functionality from these class methods
Now it is only going to be used in the `index.ts` file, where these helper methods will act as checks when the middleware is instantiated

jondeweydev committed:
lint

shalarewicz commented on src/middleware/APIAuth.ts:
```diff
@@ -0,0 +1,81 @@
+            .catch((err) => new Error(`[Log API] Server not running ${err}`));
+
+        // if project query is the wrong length: ?project=[projectID] is required
+        if (this.projectID?.length !== 24)
```
Let's avoid magic numbers. Can you create global variables for string lengths on line 45 and 49

shalarewicz reviewed (approved):
Makes sense to me. But probably have the same questions as Evan so let's chat at some point.

Only comment is on the magic numbers being used which might get lost later. Otherwise looks good!

jondeweydev committed:
updated tests and middleware with constants for easier changes if needed later on

jondeweydev committed:
merge conflicts

jondeweydev committed:
package lock fixed

evanmcneely commented on src/middleware/APIAuth.ts:
```diff
@@ -0,0 +1,81 @@
+    private logKey: string;
+
+    constructor(gateURI: string, projectID: string, logKey: string) {
+        this.gateURI = gateURI;
```
I don't know why `gateURI` has to be passed in through constructor when it is the same for every implementation of logger. It could just be hard coded and eliminate the chance of developer error.

evanmcneely reviewed (approved):
I'm approving this as a good base to work off of but with to discuss this in more detail as we proceed.
